torch/distributed/fsdp/_init_utils.py

```diff
@@ -533,7 +539,7 @@ def _get_ignored_modules(
         if isinstance(module, fsdp_file.FullyShardedDataParallel):
             raise ValueError("`ignored_modules` should not include FSDP modules")
     # Include child modules and exclude nested FSDP modules themselves
-    ignored_modules = set(
+    ignored_modules.update(
         child
         for module in ignored_root_modules
         for child in module.modules()
```

Comment:
Just to make sure this is expected:  if `_ignored_modules ` contains a `fully_shard` child, the subtree rooted at `fully_shard` will be added to `ignored_modules`. However, if `_ignored_modules` is empty or None, only the `fully_shard` module will be added?

Reply:
> if `_ignored_modules`  contains a `fully_shard` child, the subtree rooted at `fully_shard` will be added to `ignored_modules`.

Yes.

> However, if `_ignored_modules` is empty or `None`, only the `fully_shard` module will be added?

I am not sure why the `fully_shard` module would be added to `ignored_modules`? Should it?

---

I also reworked the logic to ignore modules since your comment (though I do not think it affected your questions). The semantics now are:
- Passing `ignored_modules=None` is equivalent to passing `ignored_modules=[]`.
- FSDP converts the ignored modules into a set to de-dup.
- FSDP adds any modules that cannot compose with `fully_shard` to the set. Note that in the wrapper path, we do not expect any `replicate` annotations, so this step should be a no-op.
- The remaining logic is as before. We recurse over all module subtrees for the modules in the ignored modules set (including those annotated with `replicate` for example) and ignore the entire subtree. One caveat is that `FullyShardedDataParallel` modules are not included; this is from existing logic. I think we should still include nested `fully_shard`-annotated modules if they exist in an ignored module's subtree. This is how we want to compose `fully_shard` above `replicate` above `fully_shard`.
- We also include any ignored modules already computed by child `fully_shard` or `FullyShardedDataParallel` modules.

The last two steps seem to have redundancy, but we would need to explore that more in a follow-up.